File: 2023/6/sol.js
const fs = require('fs');

function ImportFile(fileName)
{
 
    let data = new Object();
    var file = fs.readFileSync(fileName, 'utf8');
    var lines = file.split("\n");

    let times = lines[0].split("Time: ")[1].replace(/\r/gm,'').split(" ").filter(x => x !== '');
    let distance = lines[1].split("Distance: ")[1].replace(/\r/gm,'').split(" ").filter(x => x !== '');


    
    data.times = times.map(x => parseInt(x));
    data.distance = distance.map(x => parseInt(x));


    return data;
}



function PartOne(input) {

    let sum = 1;
    for (let i = 0; i < input.times.length; i++)
    {
      let time = input.times[i];
      let distance = input.distance[i];
      let solution = Math.ceil(((-time) + Math.sqrt(Math.pow(time, 2) - (4 * (-1) * (-(distance + 1)))))/-2);
      let solution2 = Math.floor(((-time) - Math.sqrt(Math.pow(time, 2) - (4 * (-1) * (-(distance + 1)))))/-2);
      let total = Math.max(solution, solution2) - Math.min(solution, solution2) + 1;

      sum *= total;
    }

    return sum;
  }
  
function PartTwo(input) {
    
    //add up all the times in the array
    let time = parseInt(input.times.join(''));
    let distance = parseInt(input.distance.join(''));
    let sum = 1;
      let solution = Math.ceil(((-time) + Math.sqrt(Math.pow(time, 2) - (4 * (-1) * (-(distance + 1)))))/-2);
      let solution2 = Math.floor(((-time) - Math.sqrt(Math.pow(time, 2) - (4 * (-1) * (-(distance + 1)))))/-2);
      let total = Math.max(solution, solution2) - Math.min(solution, solution2) + 1;

      sum *= total;

    return sum;
  }


let data = ImportFile("text.txt");


console.log(PartOne(data));
console.log(PartTwo(data));







//helpers
function CleanArray(input, char)
{
  return input.filter(function (el) {
    return el !== char;
  });
}


function isNumber(char) {
  return /^\d$/.test(char);
}

File: 2015/15/sol.py
import itertools


def ImportFile(r):
    f = open(r, 'r')
    data = dict()
    for _ in f.read().split('\n'):
        data[_.split(':')[0]] = dict()
        for __ in _.split(': ')[1].split(', '):
            data[_.split(':')[0]][__.split(' ')[0]] = __.split(' ')[1]

    return data


def P1(data, diet=False):
    perms = list(itertools.combinations_with_replacement(data.keys(), 100))
    if diet:
        perms = Diet(data, perms)
    keys = list(data.keys())
    max = 0
    for _ in perms:
        counts = dict()
        total = 1
        for i in range(len(keys)):
            counts[keys[i]] = _.count(keys[i])

        for ingrdnt in data[keys[i]].keys():
            if ingrdnt == 'calories':
                continue
            sum = 0
            for l in counts.keys():
                sum += int(data[l][ingrdnt]) * counts[l]
            if sum > 0:
                total *= sum
            else:
                total *= 0
        if total > max:
            max = total

    return max


def Diet(data, perms):
    keys = list(data.keys())
    r_perms = []
    for _ in perms:
        counts = dict()
        for i in range(len(keys)):
            counts[keys[i]] = _.count(keys[i])
        calories = 0
        for i in counts.keys():
            calories += int(data[i]['calories']) * counts[i]

        if calories == 500:
            r_perms.append(_)
    return r_perms


data = ImportFile('text.txt')
print(P1(data))
print(P1(data, True))
